help cmd | man <image>

ArangoGutierrez opened the pull request:
**Description of the Pull Request (PR):**

Add support for help cmd
```bash
$ cat my_container.def
Bootstrap: docker
From: busybox

%help
    Some help for this container

$ sudo singularity build my_container.sif my_container.def
Using container recipe deffile: my_container.def
[...snip...]
Cleaning up...

$ singularity help my_container.sif


    Some help for this container
```

**Checkoff for all PRs:**

- [x] I have read the [Guidelines for Contributing](https://github.com/singularityware/singularity/blob/master/CONTRIBUTING.md), and this PR conforms to the stated requirements.
- [ ] I have added changes to the [CHANGELOG](https://github.com/singularityware/singularity/blob/development/CHANGELOG.md) and and documentation updates to the [singularityware](https://www.github.com/singularityware/singularityware.github.io) documentation base.
- [x] I have tested this PR locally with a `make test`
- [x] This PR is against the project's master branch
- [x] I have added myself as a contributor to the [contributors's file](https://github.com/singularityware/singularity/blob/master/CONTRIBUTORS.md)
- [x] This PR is ready for review and/or merge


Attn: @singularityware-admin

ArangoGutierrez committed:
singularity help my_container.sif display help for container if available

Help should be stated on the definition file
```
Bootstrap: docker
From: busybox

%help
    Some help for this container
```

ArangoGutierrez committed:
Add HelpCmd to actionsCmds

ArangoGutierrez committed:
LICENCE => LICENSE.md

ArangoGutierrez committed:
add usage and help for HelpCmd

jmstover commented on src/docs/content.go:
```diff
@@ -616,4 +616,29 @@ Enterprise Performance Computing (EPC)`
+	// ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+	// help
+	// ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+	HelpUse   string = `help <image path>`
```
This should probably be:
```
HelpUse   string = `help <command|image path>`
```
Which would also change the other text so it isn't so focused on the help of an *image* but help in general

ArangoGutierrez replied:
good catch, thanks!

jmstover commented on src/cmd/singularity/cli/help.go:
```diff
@@ -0,0 +1,46 @@
+// HelpCmd singularity help
+var HelpCmd = &cobra.Command{
+	DisableFlagsInUseLine: true,
+	Args:   cobra.ExactArgs(1),
```
This is wrong isn't is? I mean... what if you do:
```
singularity help instance start
```
It'd only look at the help for `instance` and not `instance start`

ArangoGutierrez replied:
ummm I don't know how cobra can handle that ....

ArangoGutierrez replied:
ha!  found!

ArangoGutierrez committed:
Fix help messages content

Fix content.go to be in line with help command|image

ArangoGutierrez committed:
rm cobra.ExactArgs(1) restriction

jmstover commented on src/docs/content.go:
```diff
@@ -616,4 +613,29 @@ Enterprise Performance Computing (EPC)`
+	// help
+	// ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+	HelpUse   string = `help <command|image path>`
+	HelpShort string = `Display help for command||container if available`
```
`command||container`
Change to either `command|container` _or_ `command or container image`

ArangoGutierrez committed:
fix HelpLong and HelpExample

ArangoGutierrez committed:
Merge branch 'master' into help_cmd

ArangoGutierrez committed:
Merge branch 'master' into help_cmd

ArangoGutierrez committed:
Merge branch 'master' into help_cmd

ArangoGutierrez committed:
Merge branch 'master' into help_cmd

bauerm97 commented on src/cmd/singularity/cli/actions.go:
```diff
@@ -31,6 +31,7 @@ func init() {
 		ExecCmd,
 		ShellCmd,
 		RunCmd,
+		HelpCmd,
```
Not sure I understand why we're adding a bunch of flags to the `HelpCmd` here. We don't want to do `singularity help --bind /path container` if we just want the help message in the container.

bauerm97 commented on src/cmd/singularity/cli/help.go:
```diff
@@ -0,0 +1,50 @@
+		if _, err := os.Stat(args[0]); err == nil {
+			// Help prints (if set) the sourced %help section on the definition file
+			a := []string{"/bin/cat", "/.singularity.d/runscript.help"}
+			execStarter(cmd, args[0], a, "")
```
We should not be using `execStarter` here. We just need to set up a bare `EngineConfig`, put it into `CommonConfig`, and use `exec.Pipe`

ArangoGutierrez committed:
use exec.pipe instead of execStarter

ArangoGutierrez committed:
Merge branch 'master' of https://github.com/singularityware/singularity into help_cmd

ArangoGutierrez committed:
Merge branch 'master' into help_cmd

ArangoGutierrez committed:
Merge branch 'help_cmd' of github.com:ArangoGutierrez/singularity into help_cmd

ArangoGutierrez committed:
use run-help to avoid ambiguity

bauerm97 commented on src/cmd/singularity/cli/run_help.go:
```diff
@@ -0,0 +1,76 @@
+	PreRun:                sylabsToken,
+	Args:                  cobra.ExactArgs(1),
+	Run: func(cmd *cobra.Command, args []string) {
+
```
Remove blank line

bauerm97 commented on src/docs/content.go:
```diff
@@ -38,109 +35,109 @@ Enterprise Performance Computing (EPC)`
 
   IMAGE PATH:
-  
+
```
Are we sure we're not introducing too much whitespace here in all of these help messages? We want to make sure to be outputting in a very readable form. I'm indifferent on this, but I think @GodloveD should have good insight into if this is too much whitespace

ArangoGutierrez replied:
I'm not adding a white line here, It seems, that my IDE, just removed a [tab/space] on that line, and replaced it with a clean line.

bauerm97 replied:
Ahh okay, so it's not actually a change. No worries then it's already been approved by DG ;) carry on

bauerm97 commented on src/cmd/singularity/cli/run_help.go:
```diff
@@ -0,0 +1,76 @@
+	Args:                  cobra.ExactArgs(1),
+	Run: func(cmd *cobra.Command, args []string) {
+
+		if _, err := os.Stat(args[0]); err == nil {
```
What happens when `err != nil`? We need to handle `err` after the conditional here. Some kind of error like `container not found` or something

ArangoGutierrez replied:
ups, fixing

ArangoGutierrez replied:
Addressed on f0e95e2

ArangoGutierrez committed:
add error handling

bauerm97 commented on src/docs/content.go:
```diff
@@ -616,4 +613,30 @@ Enterprise Performance Computing (EPC)`
+	// Run-help
+	// ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+	RunHelpUse   string = `run-help <image path>`
+	RunHelpShort string = `Display help for command|container if available`
```
This command does not display help for a command.

ArangoGutierrez committed:
fix typo on RunHelpShort @ docs/content.go

ArangoGutierrez committed:
Merge branch 'master' into help_cmd

GodloveD commented on src/cmd/singularity/cli/run_help.go:
```diff
@@ -0,0 +1,77 @@
+	PreRun:                sylabsToken,
+	Args:                  cobra.ExactArgs(1),
+	Run: func(cmd *cobra.Command, args []string) {
+		if _, err := os.Stat(args[0]); err == nil {
```
Instead of having the entire `Run` section wrapped in this `if` you could do something like this and then just move on with the rest of the code:
```
if _, err := os.Stat(args[0]); err != nil {
        sylog.Fatalf("container not found: %s", err)
}
<rest of code>
```

bauerm97 replied:
Agree with this comment

GodloveD commented on src/docs/content.go:
```diff
@@ -616,4 +613,30 @@ Enterprise Performance Computing (EPC)`
+  Cleaning up...
+
+  $ singularity run-help my_container.sif
+
```
It looks like this implementation doesn't add a bunch of extra whitespace like the 2.x version did.  Should remove extra whitespace from example.

GodloveD commented on src/cmd/singularity/cli/run_help.go:
```diff
@@ -0,0 +1,77 @@
+			// Help prints (if set) the sourced %help section on the definition file
+			abspath, err := filepath.Abs(args[0])
+			name := filepath.Base(abspath)
+			a := []string{"/bin/cat", "/.singularity.d/runscript.help"}
```
Is this how the old version of help used to work.  What if there is no `/bin/cat` in the container.  No need to change anything based on this comment.  Just curious.

jmstover replied:
@GodloveD basically yeah... 2.x runs:
`eval_abort cat "${SINGULARITY_MOUNTPOINT}/.singularity.d/runscript.help"`

In `libexec/helpers/help.sh`

ArangoGutierrez replied:
Yeah I double checked with @jmstover this one

GodloveD commented on src/cmd/singularity/cli/run_help.go:
```diff
@@ -0,0 +1,77 @@
+			}
+
+			configData, err := json.Marshal(cfg)
+			if err != nil {
```
For consistency, it'd be good to combine the two lines above like so:
```
if configData, err := json.Marshal(cfg); err != nil {
```

bauerm97 replied:
That can't be done here. If we do that, then line 64 `configData` will be undefined. Variables which are declared/set in an if statement are scoped:

```
if test := 1; test == 1 {
    fmt.Println(test) // Prints "1"
}

fmt.Println(test) // Compiler error
```

The way Ed did it here is the idiomatic handling

GodloveD replied:
I stand corrected.  Thanks @bauerm97.

GodloveD reviewed (dismissed):
Just a bunch of style nitpicks.  

So I guess this is the basic set of things that must be done to mount a container?!  Seems like a lot!  

¯\_(ツ)_/¯

ArangoGutierrez committed:
trim whitelines

ArangoGutierrez committed:
better err handling

Instead of having the entire Run section wrapped in a big "if"
do a better err check and the the rest of the code:

```
if _, err := os.Stat(args[0]); err != nil {
        sylog.Fatalf("container not found: %s", err)
	}
	<rest of code>
```

ArangoGutierrez committed:
Merge branch 'help_cmd' of github.com:ArangoGutierrez/singularity into help_cmd

bauerm97 committed:
Merge branch 'master' into help_cmd
